backend/pcloud/chunkwriter.go

```diff
@@ -0,0 +1,337 @@
+	// In case the file existed previosly and was larger than the newly written file, we need to truncate it to the right size.
+	byteCount := c.GetBytesCount()
+	fs.Debugf(c.src, "truncate target to %d bytes", byteCount)
+	if _, err := c.fileTruncate(ctx, byteCount); err != nil {
+		return fmt.Errorf("truncate file: %w", err)
+	}
+
+	// close fd
+	if _, err := c.fileClose(ctx); err != nil {
+		return fmt.Errorf("close fd: %w", err)
+	}
+
+	// There seems to be no option in the fileops API to set the modtime.
+	// Therefore, we apply the modtime via the regular file API.
+	o, err := c.fs.NewObject(ctx, c.remote)
+	if err != nil {
+		return fmt.Errorf("open newly copied object: %w", err)
+	}
+	if err := o.SetModTime(ctx, c.src.ModTime(ctx)); err != nil {
+		return fmt.Errorf("set modtime: %w", err)
+	}
+
+	return nil
+}
+
+// WriteChunk implements fs.ChunkWriter.
+func (c *chunkWriter) WriteChunk(ctx context.Context, chunkNumber int, reader io.ReadSeeker) (bytesWritten int64, err error) {
+	return c.writeChunk(ctx, chunkNumber, reader)
+}
+
+func (c *chunkWriter) writeChunk(ctx context.Context, chunkNumber int, reader io.Reader) (bytesWritten int64, err error) {
+	chunk := int64(chunkNumber)
+
+	// read chunk into buffer
+	buffer := make([]byte, c.chunkSize)
+	offset := chunk * c.chunkSize
+	eg, egCtx := errgroup.WithContext(ctx)
+
+	// read input buffer and calculate hash
+	inSHA1 := ""
+	eg.Go(func() error {
+		byteCount, err := io.ReadFull(reader, buffer)
+		buffer = buffer[:byteCount]
+		if err != nil && !errors.Is(err, io.ErrUnexpectedEOF) && !errors.Is(err, io.EOF) {
+			return err
+		}
+		inSHA1Bytes := sha1.Sum(buffer)
+		inSHA1 = hex.EncodeToString(inSHA1Bytes[:])
+		return nil
+	})
+
+	// get target hash
+	outSHA1 := ""
+	eg.Go(func() error {
+		targetChecksum, err := c.fileChecksum(egCtx, offset, c.chunkSize)
+		if err != nil {
+			return err
+		}
+		outSHA1 = targetChecksum.SHA1
+		return nil
+	})
+
+	if errors.Is(err, io.EOF) {
+		return 0, nil
+	}
+
+	if err := eg.Wait(); err != nil {
+		return 0, fmt.Errorf("check hashes of chunk %d: %w", chunk, err)
+	}
+	if outSHA1 == "" || inSHA1 == "" {
+		return 0, fmt.Errorf("block %d: expect both hashes to be filled: src: %q, target: %q", chunk, inSHA1, outSHA1)
+	}
+
+	byteCount := int64(len(buffer))
+	c.AddBytesCount(byteCount)
+	// check hash of chunk, skip if fits
+	if inSHA1 == outSHA1 {
+		fs.Debugf(c.src, "chunk %d matches", chunk)
+		return byteCount, nil
+	}
+
+	// upload buffered chunk with offset if necessary
+	if _, err := c.filePWrite(ctx, offset, buffer); err != nil {
+		return 0, err
+	}
+	fs.Debugf(c.src, "chunk %d written", chunk)
+	return byteCount, nil
+}
+
+// Call pcloud file_open using folderid and name with O_CREAT and O_WRITE flags, see [API Doc.]
+// [API Doc]: https://docs.pcloud.com/methods/fileops/file_open.html
+func fileOpenNew(ctx context.Context, c *rest.Client, srcFs *Fs, directoryID, filename string) (*api.FileOpenResponse, error) {
```

Comment:
These file operations look very general.

In fact general enough to implement the `OpenWriterAt` interface

```
	// OpenWriterAt opens with a handle for random access writes
	//
	// Pass in the remote desired and the size if known.
	//
	// It truncates any existing object
	OpenWriterAt func(ctx context.Context, remote string, size int64) (WriterAtCloser, error)
```

That is a more general interface than the `OpenChunkWriter` .

Is it possible to implement this? Rclone will use this in preference to the `OpenChunkWriter` interface for multithread uploads.

What do you think?